src/discof/forest/fd_forest.c

```diff
@@ -461,22 +463,80 @@ fd_forest_publish( fd_forest_t * forest, ulong new_root_slot ) {
+        tail->prev = null;
       }
       child = fd_forest_pool_ele( pool, child->sibling );
     }
-    fd_forest_ele_t * next = fd_forest_pool_ele( pool, head->next ); /* FIFO pop */
+    fd_forest_ele_t * next = fd_forest_pool_ele( pool, head->prev ); /* FIFO pop */
     fd_forest_pool_ele_release( pool, head ); /* free head */
     head = next;
   }
 
+  /* If there is nothing on the frontier, we have hit an edge case
+     during catching up where all of our frontiers were < the new root.
+     In that case we need to continue repairing from the new root, so
+     add it to the frontier. */
+
+  if( FD_UNLIKELY( fd_forest_frontier_iter_done( fd_forest_frontier_iter_init( frontier, pool ), frontier, pool ) ) ) {
+    fd_forest_ele_t * remove = fd_forest_ancestry_ele_remove( ancestry, &new_root_ele->slot, NULL, pool );
+    if( FD_UNLIKELY( !remove ) ) {
+      /* Very rare case where during second incremental load we could publish to an orphaned slot */
+      remove = fd_forest_orphaned_ele_remove( orphaned, &new_root_ele->slot, NULL, pool );
+    }
+    FD_TEST( remove == new_root_ele );
+    fd_forest_frontier_ele_insert( frontier, new_root_ele, pool );
+    new_root_ele->complete_idx = 0;
+    new_root_ele->buffered_idx = 0;
+    advance_frontier( forest, new_root_ele->slot, 0 );
+  }
+
   new_root_ele->parent = null; /* unlink new root from parent */
-  forest->root     = fd_forest_ancestry_idx_query( ancestry, &new_root_slot, null, pool );
+  forest->root         = fd_forest_pool_idx( pool, new_root_ele );
+
+  /* Lastly, cleanup orphans if there orphan heads < new_root_slot.
+     First, add any relevant orphans to the prune queue. */
+
+  head = NULL;
+  for( fd_forest_orphaned_iter_t iter = fd_forest_orphaned_iter_init( orphaned, pool );
+       !fd_forest_orphaned_iter_done( iter, orphaned, pool );
+       iter = fd_forest_orphaned_iter_next( iter, orphaned, pool ) ) {
+    fd_forest_ele_t * ele = fd_forest_orphaned_iter_ele( iter, orphaned, pool );
+    if( FD_UNLIKELY( ele->slot < new_root_slot ) ) {
+      if( FD_UNLIKELY( !head ) ) {
+        head = ele;
+        head->prev = null;
+        tail = ele;
+      } else {
+        tail->prev = iter.ele_idx;
+        tail = fd_forest_pool_ele( pool, tail->prev );
+        tail->prev = null;
+      }
+    }
+  }
+
+  /* Now BFS and clean up children of these orphan heads */
+  while( head ) {
+    fd_forest_ele_t * child = fd_forest_pool_ele( pool, head->child );
+    while( FD_LIKELY( child ) ) {
+      if( FD_LIKELY( child != new_root_ele ) ) {
+        tail->prev = fd_forest_pool_idx( pool, child ); /* insert prune queue */
+        tail       = fd_forest_pool_ele( pool, tail->prev ); /* advance prune queue */
+        tail->prev = null;
+      }
+      child = fd_forest_pool_ele( pool, child->sibling );
+    }
+    ulong remove = fd_forest_orphaned_idx_remove( orphaned, &head->slot, null, pool ); /* remove myself */
```

Comment:
clear `remove->next` too in case there was a map collision (can lead to subtle bugs)
```suggestion
    fd_forest_ele_t * remove = fd_forest_orphaned_ele_remove( orphaned, &head->slot, null, pool );
    remove->next = idx_null;
```